### ISSUE TITLE
Control Net Input for Rigging

### ISSUE THREAD

**Issue by ForestEco** [opentoonz/opentoonz#5445]
### Is your feature request related to a problem? Please describe

I have a Python script that will process a stack of images to generate a control net. If someone is inspired by precise movement of a character, then a control net might be useful. 

### Describe the solution you'd like

Perhaps adapt my python script for C++ to analyze multiple photos and save control nets and generate an input for the rigging system.

**Comment by RodneyBaker** [comment]
@ForestEco 
I believe we can assume this is the python script you reference:  https://github.com/ForestEco/Stable_Diffusion_Py
If that isn't the case please let us know.

**Comment by ForestEco** [comment]
@RodneyBaker 
Yes, that's right. I adapted the script to work in batches. Feel free to use it, or perhaps I can help. I know C#, but not C++, so I am not sure how much I offer now. I've got a C++ game design course to start and I know Paint.net might be worth exploring in C++.

**Comment by RodneyBaker** [comment]
Where it comes to AI (control net and the like) it would seem python is about as good a place to focus as any at the moment.
Most graphics programs don't require full integation of python in order for that to be useful.

**Comment by ForestEco** [comment]
Ok, so you will be able to use that so long as the licensing it is derived from is good to go? Do you need more support with Python programming or developing OT plugins? I had dove deeply into machine learning and ComputerVision topics in the last year for Python. I learned about FiftyOne which manages stacks of photos for using Segment Anything Model to find specific features in photostock, so I made a module, which was able to run yet my PC was too slow, I needed to rent a GPU because some of this stuff is still in the future for OS/File Explorer. 

It would be cool to learn how to edit the way OT works on the canvas level or run models to make the tools work or customize some. What modules would I need to learn?

**Comment by RodneyBaker** [comment]
There currently is no direct onramp for python in Opentoonz.
Perhaps the closest that exists is OpenCV and some FX does use machine learning (the example here might be the Neural Style Transfer FX node that no one in the wild has demonstrated using).  It does however represent one of the early attemps at adding machine learning to animation programs prior to the current surge of AI popularity.

> Ok, so you will be able to use that so long as the licensing it is derived from is good to go? 

It's hard to say what can be used.  
We are mostly a community of user-developers... those that see a need and implement what they can.


> Do you need more support with Python programming or developing OT plugins?

We defiinitely do need development of connectivity with python and OT plugins.

All of this is why I suggest developing python support in parallel with OT development.
Python doesn't need actual connectivity to OT in order to be useful; an example being your script which just points to a directory of images.  Opentoonz might render the images... a script such as yours process those images... conversion or plugin use then is applied to that result to further utilize those resources.

It will take a lot of development for sure.

> What modules would I need to learn?

The two onramps that immediately come to mind would be Shaders and OPenCV plugins.
Both of these use the node inside the FX Schematic to present information and options for user to set in the flow of operations from initial input to final output.

The most difficult part of Shader development (in my estimation) is not so much the shaders themselves but the difficulty in installing/activating those Shaders in Opentoonz.  There are about 5 files that either need to be created and properly placed or edited to contain appropriate entries for Opentoonz to recognize they exist and be available to the user.  This process needs to be refined so that users don't have to go through that complex process to add Shaders.

I mention Shaders because the use of OpenCV plugins is very similar and uses many of the same resources... and the plugins present themselves in much the same way to users inside the FX Schematic.

Here's a link to the Neural Style Plugin's repo as a starting place.  Note that no update to that code has been made in 7 to 8 years... basically since Opentoonz and related resources were first open sourced.
Link:  https://github.com/opentoonz/neural_style_plugin

I have heard that others have had a working means to add image generative prompts into the FX Schematic but don't have any personal knowledge of that.  @konero might be able to point you in the right direction there as well as suggest tips on how best to integrate such developments into Opentoonz.  His focus is mainly on the UI and stylesheets but understands some of the complexity around that quite well.

**Comment by ForestEco** [comment]
Thanks for the thorough response and resources to start learning. I am familiar with ImageJ library, and it is all similar to CV. So developing special effects sounds right on ramp for me, but it can be difficult to get them to render in OT you say? I will have to explore a bit.

**Comment by ForestEco** [comment]
One more question for you @RodneyBaker, I am trying to use the Fx and limiting it to certain frames but the keyframe tool seems to be applying it to any keyframes. Is there a tutorial that explains well how to limit the Fx to specific key frames or do I have to make another scene and just pick those frames to splice? Is there a course on using the FX?

**Comment by RodneyBaker** [comment]
I'm a fan of ImageJ and its various 'ports'.  Haven't used it much but some of it's toolings and processes are fairly unique and dare I say powerful.  Very cool that you are familiar with it.  The Java side of that equation though makes it perhaps yet another track in the many different paths that run parallel to C++ and only occassionally find someone with the drive to bridge gaps.  Like much of current AI generative output, most is very compatible with producing content with Opentoonz.  In that sense... an image is an image is an image (or sequence of images).  As such the output of one path can be the input of another and bridge many an incompatibility.

Regarding use of FX and keyframes.
Each FX node has it's own keyframes which can be set per variable/setting.
There is a Schematic viewer where we can click on the diamond shaped keyframe icon and key all settings of that FX per frame but usually we just set individual keys for each setting.

That is via the FX Schematic.
Using the Function Edtior (and Curve Editor) we get different views of these same keys and channels.
Is there a tutorial that walks us through all of that?  Hmmm... not to the depths that I'd like to see.
The complexity of what can be done is sufficient enough that it can be better to state a desired outcome and then have a custom tutorial created to explore various means to get there.

We haven't done it recently but there has been interest in folks joining live sessions where Q&A is presented and explored.
That approach might be of more use to those whose tolerance for complexity is higher than most.  In other words, those more likely to crack open the options of the Function Edtior, Curve Editor... those interested in diving deep into how to not just dig deep into FX but to build their own.

Having said that, here are some of the youtube videos on keyframing (some more appropriate than others):

https://www.youtube.com/results?search_query=opentoonz+keyframe+Function+Editor

**Comment by ForestEco** [comment]
> Since the tech is proven to work, something must have generated that proof of it working, right? Meaning, theres something out there that can use these control nets and rig images with it.
> 
> What would opentoonz do better than the tools already out there that were already able to do it...already? Or rather, if opentoonz cannot do the process better, what would it offer to work with these rigs?

What software are you specifically mentioning? 
My thought is that the control nets may be useful in the case rigging is difficult, such as getting the natural movements of some fast moving wildlife for example or some intricate expert poses that are not so easy to replicate using the human body. Such as modeling a superhero after a famous athlete's choice of movement. What I was hoping to do is get OT to build a skeleton for rigging off the control net, then fit the skeleton to other columns or whatever you want to develop the rig. That way you get the best possible line tracing as you go. But you might just use my script to like you said, just make a reference image and move the rig to it. Problem solved?
